src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -282,58 +283,50 @@ class SegmentedAllocator : public ExtentAllocator {
   std::vector<Writer> writers;
   LBAManager& lba_manager;
   Journal& journal;
-  Cache& cache;
 };
 
 class ExtentPlacementManager {
 public:
   ExtentPlacementManager(
-    Cache& cache,
     LBAManager& lba_manager
-  ) : cache(cache), lba_manager(lba_manager) {}
+  ) : lba_manager(lba_manager) {}
 
-  /**
-   * alloc_new_extent_by_type
-   *
-   * Create a new extent, CachedExtent::poffset may not be set
-   * if a delayed allocation is needed.
-   */
-  CachedExtentRef alloc_new_extent_by_type(
+  struct alloc_result_t {
+    paddr_t paddr;
+    bufferptr bp;
+  };
+  alloc_result_t alloc_new_extent(
```

Comment:
Since all other methods of EPM are operating CachedExtent, I think maybe we should move the creation of CachedExtent from `Cache::alloc_new_extent` into `ExtentPlacementManager::alloc_new_extent`, so that we have a consistent EPM interface.

Reply:
I think maybe the `alloc_new_extent` function can be entirely moved into EPM, it seems that extent allocation should be the responsibility of EPM and `Cache` doesn't have to possess the extent allocation power, am I right?

Reply:
IIUC `EPM` only needs to worry about extent placements, currently in two scenarios:
A. Decide during cached extent allocation with `Cache`;
B. Decide during the transaction phase `write_pipeline.ool_writes` for delayed decisions from A;

In scenario A, I think `EPM` is appropriate to work behind `Cache` as it essentially needs to make two decisions:
1. paddr:
    a. Whether the decision of paddr should be delayed until scenario B;
    b. If no delay, whether the paddr is absolute (for future RBM and PMem devices) or record-relative (for Segmented devices);
2. buffer -- for PMem, the buffer is possible to point to the memory address within PMem directly;

In addition, `Cache` exposes two heterogeneous interfaces `alloc_new_extent<T>()` and `alloc_new_extent_by_type()` for the dependent components to make allocations. `EPM` can have a uniform interface `EPM::alloc_new_extent(type, hint, len) -> paddr, buf` to support the scenario A if work behind `Cache`.

For scenario B, I think there are two directions, achitecture-wise:
B1. Make `EPM` more complicated to coordinate the `write_pipeline.ool_writes` phase;
B2. Let `TransactionManager` to coordinate the `write_pipeline.ool_writes` phase, and try simplify the responsibilities of `EPM` with less dependencies;

I believe B2 is a more appropriate direction because `TransactionManager` is already the coordinator of various components `Cache`, `Transaction`, `LBAManager`, etc. It seems to me is not necessary for `EPM` to become another coordinator that require such dependencies.

As for https://github.com/ceph/ceph/pull/44809#discussion_r800288013, I didn't have the chance to dig into this yet, but IIUC `TransactionManager` can still be the coordinator to process extent split/merge with dependent components, with or without `EPM`, under proper further refactors. Does it make sense?

Reply:
> In scenario A, I think `EPM` is appropriate to work behind `Cache` as it essentially needs to make two decisions:
> 
> 1. paddr:
>    a. Whether the decision of paddr should be delayed until scenario B;
>    b. If no delay, whether the paddr is absolute (for future RBM and PMem devices) or record-relative (for Segmented devices);
> 2. extent buffer -- for PMem, the buffer is possible to point to the memory address within PMem directly;
> 
> In addition, `Cache` exposes two heterogeneous interfaces `alloc_new_extent<T>()` and `alloc_new_extent_by_type()` for the dependent components to make allocations. `EPM` can have a uniform interface `EPM::alloc_new_extent(type, hint, len) -> paddr, buf` to support the scenario A if work behind `Cache`.

You are right, but what I meant `Cache` is not a better place to take on the extent allocation responsibility than EPM. I just don't think it's necessary to split the allocation functionality. Am I right?



> As for [#44809 (comment)](https://github.com/ceph/ceph/pull/44809#discussion_r800288013), I didn't have the chance to dig into this yet, but IIUC `TransactionManager` can still be the coordinator to process extent split/merge with dependent components, with or without `EPM`, under proper further refactors. Does it make sense?

Splitting/merging extents involves extent placement, so I think EPM's delayed allocation is the last chance to do the job.

Reply:
> In addition, `Cache` exposes two heterogeneous interfaces `alloc_new_extent<T>()` and `alloc_new_extent_by_type()` for the dependent components to make allocations

The return value of the two interfaces are actually both cached extents.

Reply:
> I just don't think it's necessary to split the allocation functionality into two parts, we can just put the whole allocation function into EPM, this can simplify the implementation even more.

LBAtree also relies on cache for allocations, we will need to adjust more code if put EPM in front of cache.

Reply:
> > I just don't think it's necessary to split the allocation functionality into two parts, we can just put the whole allocation function into EPM, this can simplify the implementation even more.
> 
> LBAtree also relies on cache for allocations, we will need to adjust more code if put EPM in front of cache.

I didn't mean to put EPM in front of cache, I meant we use EPM to replace cache as the extent allocator. And I think simplifying implementations doesn't necessarily mean adjusting less code, as long as we can have a clear and simple result, it's worth it, am I right?

Reply:
> I meant we use EPM to replace cache as the extent allocator.

I don't really see the benefit here, this still needs `EPM` to add the allocated extents into `Cache`, right? I think it is semantically similar with the current approach for `Cache` to ask for `EPM`'s decision when allocating new extents. And LBA tree also relies on other interfaces of `Cache`.

Reply:
> I don't really see the benefit here, this still needs `EPM` to add the allocated extents into `Cache`

This is already done.

> And LBA tree also relies on other interfaces of `Cache`.

The only use of other interfaces of `Cache` is `Cache::get_extent()`. Extents should be retrieved from cache, but don't necessarily get allocated by cache. I think extents allocations' major job is choose a right place to put it, so EPM should be a better place. The benefit here is that we have a clear extent allocation interface, and since we already put EPM in front of cache, this approach adjusts less code.

I think simplifying the implementation doesn't necessarily mean referencing less components. Take SegmentCleaner and SegmentManager as an example, if we only consider referencing less components, those two should be merged into one, right? I think the bottom line of simplifying the implementation is to have a clear and simple component function boundary. But these are just my thoughts.

Reply:
Hmm, @athanatos what do think about the commit https://github.com/ceph/ceph/pull/44809/commits/405e7bc2d44e1e673c4f8d98c5ad987b2437be74 ?

Reply:
> I still don't see "whether `EPM` should behind `Cache` or not" is a big issue.

It's not about whether `EPM` should be in front of `Cache` or behind it. It's about how the extent allocation interface should be. And if "whether `EPM` should behind `Cache` or not" is not a big issue, why change it from EPM in front of `Cache` to behind it, in the first place?

Reply:
I actually think this is fine.  The patch removes ExtentPlacementManager's Cache reference the delayed argument in Cache was a bit duplicative of the EPM's allocation implementation.

Reply:
I'm not sure what you mean by splitting/merging extents?

Reply:
> I'm not sure what you mean by splitting/merging extents?

RIght now, if we have a 64KB data extent and modify a relative small portion, say 4 kb of it, the whole 64kb extent will be read out and rewritten, although this is functionally fine, it is very inefficient. we need to be able to split that 64kb extent into multiple smaller ones.

On the other hand, in scenarios where hard disks are used and we need "min_alloc_unit" to prevent the hard disks from being over fragmented, we also need that extent split capability to deal with scenarios where small overwrites that aiming at large extents lying in the hard disks need to be stored on faster devices.

There are circumstances where small extents that are logical-address-adjacent need to be merged into a large extent so that logically sequential reads won't be mapped into multiple random reads.

Reply:
> the delayed argument in Cache was a bit duplicative of the EPM's allocation implementation.

I think we can also remove the delayed argument in Cache and the Cache reference in EPM, if we put the whole extent allocation function into EPM. And what's more, this approach will avoid the need to put EPM reference into Cache. Am I right?

Reply:
From our discussion in the weekly meeting, it seems like the issue is that split/merge would be a problem if it needs special handling during the EPM's delayed handling pathway.  I think we're going to go with this PR for now and try to structure split/merge so that it doesn't need special handling outside of ObjectDataHandler.  @xxhdx1985126 Does that seem ok?

Reply:
> From our discussion in the weekly meeting, it seems like the issue is that split/merge would be a problem if it needs special handling during the EPM's delayed handling pathway. I think we're going to go with this PR for now and try to structure split/merge so that it doesn't need special handling outside of ObjectDataHandler. @xxhdx1985126 Does that seem ok?

Yeah, that's fine:-)

Reply:
Thanks for the extended discussions! Sorry missed the standup due to personal conflicts.